Lib/test/test_importlib/identity.py

```diff
@@ -0,0 +1,513 @@
+        type_candidates = [
+            ident_addr_domain_other,
+            ident_addr_local_other,
+            ident_name_other,
+            name_value_other,
+        ]
+        stop = False
+        i = len(entries) - 1
+        while i >= 0 and not stop:
+            j = len(entries[i]) - 1
+            while j >= 0 and not (stop := entries[i][j].category is quote):
+                token = entries[i][j]
+                if token.category not in type_candidates:
+                    repls = ()
+                elif token.category is ident_name_other:
+                    final = entries[i][j + 1] is ident_name_only
+                    repls = get_name_qchar_idxs(token.value, not final)
+                else:
+                    repls = range(len(token.value))
+                for k in repls:
+                    index_candidates.append((i, j, k))
+                j -= 1
+            i -= 1
+        return index_candidates
+
+    def unbalance(entries):
+        """
+        Unbalance quote characters in the list of entries without altering
+        the parsing of the list.
+        """
+        qchrs = draw(strategies.sets(strategies.sampled_from("\"'")))
+        index_candidates = unbalance_indexes(entries)
+        if not qchrs or not index_candidates:
+            return False
+        for qchr in qchrs:
+            idx_entry, idx_token, idx_char = draw(
+                strategies.sampled_from(index_candidates)
+            )
+            token = entries[idx_entry][idx_token]
+            token.value = replace(token.value, idx_char, qchr)
+        return True
+
+    # Non-leaf rules for "Author" and "Maintainer" fields:
+
+    def name_entry():
+        choices = (name_empty, name_value)
+        return draw(strategies.sampled_from(choices))()
+
+    def name_empty():
+        return Entry.from_empty(2, name_empty)
+
+    def name_value():
+        choices = (name_value_other, quote)
+        functions = draw(
+            strategies.lists(strategies.sampled_from(choices), min_size=1, max_size=10)
+        )
+        entry = Entry.from_empty(2, name_value)
+        entry[:1] = apply_merge(functions, name_value_other, sub_entry)
+        return entry
+
+    # Non-leaf rules for "Author-email" and "Maintainer-email" fields:
+
+    def ident_entry():
+        choices = (ident_empty, ident_name_only, ident_addr_only, ident_name_addr)
+        return draw(strategies.sampled_from(choices))()
+
+    def ident_empty():
+        return Entry.from_empty(2, ident_empty)
+
+    def ident_name_only():
+        entry = Entry.from_empty(2, ident_name_only)
+        entry[:1] = ident_name(name_only=True)
+        return entry
+
+    def ident_addr_only():
+        entry = Entry.from_empty(2, ident_addr_only)
+        addr = ident_addr()
+        if choose_bool():
+            delim = ident_name_addr_delim("")
+            addr = [delim, *addr]
+            entry[1:] = addr
+        else:
+            entry[2:] = addr
+        return entry
```

Comment:
As a function returning a strategy, this would look something like:

```python
def ident_empty():
    return st.just(Entry.from_empty(length=2, category="ident_empty"))

@st.composite
def ident_name_only():
    entry = Entry.from_empty(length=2, category="ident_name_only")
    entry[:1] = draw(ident_name(name_only=True))
    return entry

@st.composite
def ident_addr_only(draw):
    entry = Entry.from_empty(length=2, category="ident_addr_only")
    addr = draw(ident_addr())
    if draw(st.booleans()):
        entry[1:] = [draw(ident_name_addr_delim("")), *addr]
    else:
        entry[2:] = addr
    return entry
```

I'd also suggest moving them out of the top-level composite function, so that you don't have to redefine and re-validate everything again each time Hypothesis generates another test case.

Reply:
Oh, that's what `st.just` is for... Thanks for the tip.